JavaScript/testing.md

```diff
@@ -0,0 +1,109 @@
+
+**2. Business Requirements**
+
+When a flow is implemented, it usually determined by the requirements on the ticket itself. Sometimes these requirements are not explicitly written down or not followed to a T.
+
+**3. Data Flow**
+
+Coding is really just manipulating data. And on the front end, it's mostly just manipulating data that we either get from the backend or manipulating data structures that we create and store ourselves. Certain data problems that are fairly common are trying to manipulate data that is actually not the type we expect and trying to operate on data that is not defined.
+
+**4. Accidental Code Changes**
+
+Little hiccups get by us and others reviewing our code. We add a console.log or debugger, or even just change one line elsewhere in the application to make our local development easier and it somehow makes its way into production.
+
+_So how do we address these bugs and more importantly prevent them? Let's look at what tools and methodologies are available to us._
+
+<a name="testing-methodologies"></a>
+
+## Testing Methodologies
+
+Let's take a step back and look at what bugs we're trying to solve and prevent.
+
+<a name="testing-for-style-bugs"></a>
+
+### Testing for Style Bugs
+
+There are a few ways we can address style bugs.
+
+**1. Automated**
+
+Automated visual testing is possible, but it is definitely difficult and costly. There are products out there that will run and build your tests, click on elements, and make sure that the screenshots that are taken match previous screenshots. This is usually very helpful for regressions, but not for seeing if implementations match designs to spec.
+
+**2. Manual Through a Person**
+
+In a typical review process, the code and its subsequent implementation go through a few different hands. One important thing we can do is make sure that it gets approval by either a designer or someone with design thinking a keen sense of whitespace and type. This is typically lower cost and less upkeep. If added as a part of the review process before code is even reviewed, it can act as a great means of QA, since design can let development know that there are style bugs / inconsistencies before it is even merged. Another person that can act as an enforcer here is the other developer reviewing the code. We can make it mandatory that the developer pull the branch down, interact with it, and ensure that the style meets that of the design.
+
+**3. Onion Skinning**
+
+Onion skinning is when the design is physically placed on top of the local instance of the feature itself, often done through a tool like Zeplin, which allows the transparency of an overlay to be changed. The limitation of onion skinning is that the implementation on other phones besides the one the mockup was designed for can look bad. This leads us to **4.**
+
+**4. Design and Develop on Multiple Screens**
```

Comment:
A nod to both iOS and Android or cross-browser testing here could be good too

Reply:
good suggestion 👍